### ISSUE TITLE
[HOLD for payment 2022-07-25] [$500] Do not resize small attachments displayed in Preview Modal

### ISSUE THREAD

**Issue by thienlnam** [Expensify/App#8471]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Break down in numbered steps

Example file
![82ca9493622eeeb4d6648d0a3369e41fb8f40349_128](https://user-images.githubusercontent.com/30609178/161613034-68f73dac-0734-400d-ba3b-c7068995598f.jpeg)

1. Upload a picture with small dimensions
2. Notice that it gets resized
<img width="863" alt="Screen Shot 2022-04-04 at 11 59 24 AM" src="https://user-images.githubusercontent.com/30609178/161613159-9a0905ff-1de3-401e-b74d-869c9bb936c3.png">

## Expected Result:
A picture with a small dimensions should be displayed exactly as it is provided in the preview modal

## Actual Result:
The picture is enlarged and results in a blurry picture in the preview modal

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platform:

Where is this issue occurring?

- Web x
- iOS x 
- Android x
- Desktop App x
- Mobile Web x

**Version Number:** 
**Reproducible in staging?:** 
**Reproducible in production?:** 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:** 
**Issue reported by:** 
**Slack conversation:** 

**Job Post** https://www.upwork.com/jobs/~01079bbed339023db8

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by metehanozyurt** [comment]
Hi , in this [issue:8119](https://github.com/Expensify/App/issues/8119) my proposel fix this issue too. please review the video thank you.

https://user-images.githubusercontent.com/6093207/161835598-256de6ca-21fe-4e8f-b82d-d8414f9a21ec.mp4

**Comment by flodnv** [comment]
>Hi , in this issue:8119 my proposel fix this issue too

cc @mananjadhav ^

**Comment by parasharrajat** [comment]
Thanks, Florent. I will test the proposal shortly.

**Comment by parasharrajat** [comment]
Ok, I see your proposal @metehanozyurt But it has a lot of unneeded information for this issue like zooming and other stuff. Could you please add a minimal proposal here that is needed to fix it?  We can start with fixing this one and later those changes can be used in another issue.

**Comment by metehanozyurt** [comment]
Sorry to associate it with another more complex issue @parasharrajat 🙏.


For Web version we should change this file [ImageView/index.js](https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageView/index.js);

Step 1: 
We should change this line because when containerHeight is not yet available when the setImageRegion function is called.

this line 
https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageView/index.js#L138

change like this

```JS
if (imageHeight <= 0) {
```

Step 2:
Small size images called so fast then container layloutChanged cant get imgHeight and imgWidth from state

https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageView/index.js#L269
Should change like this

```js
onLayout={e => setTimeout(() => {
                    this.onContainerLayoutChanged(e);
                }, 1)}
``` 

Step 3:
this need change like this because on mWeb using this code block;
https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageView/index.js#L245-L255

Should change like this

```JS
                <View
                    ref={el => this.scrollableRef = el}
                    style={[styles.imageViewContainer, styles.overflowHidden]}
                    onLayout={e => setTimeout(() => {
                        this.onContainerLayoutChanged(e);
                    }, 1)}
                >
                    <Image
                        source={{uri: this.props.url}}
                        style={[
                            styles.w100,
                            styles.h100,
                        ]}
                        resizeMode={this.state.zoomScale >= 1 ? 'center' : 'contain'}
                        onLoadStart={this.imageLoadingStart}
```
Step 4:
componentDidMount should not return immediately when mWeb

https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageView/index.js#L50-L56

should change like this

```js
componentDidMount() {
        Image.getSize(this.props.url, (width, height) => {
            this.setImageRegion(width, height);
        });
        if (this.canUseTouchScreen) {
            return;
        }

```



------------------------------------------------------------------------

Step 5:
For Mobile version we should change this file [ImageView/index.native.js](https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageView/index.native.js);

we should delete this two line because this 2 line resize image.

https://github.com/Expensify/App/blob/3316204c5828132329fe8d686bb6c3aa018f2f1a/src/components/ImageView/index.native.js#L81-L82
change like this

```js                
// imageHeight *= aspectRatio;
// imageWidth *= aspectRatio;
```

**Comment by parasharrajat** [comment]
Thanks for the update proposal @metehanozyurt.

- Step 2 is prohibited in our codebase. 
- Step 3 Why is this needed? 

I don't where are you correcting the size of the image on the web.

**Comment by metehanozyurt** [comment]
Thank you @parasharrajat for checking my proposal. For step 2 i will much great suggestion🙏 or will check really need that timeout and searching workarounds. 
Step 3 needed because when i test on mWeb  I noticed that fill the picture in the view. so I had to check if the photo is smaller than the area and center it.

for web version we dont need correction image size actually step 1 can keep function continue after that this.state.zoomScale can be set.

**Comment by parasharrajat** [comment]
> Step 3 needed because when i test on mWeb I noticed that fill the picture in the view. so I had to check if the photo is smaller than the area and center it.

I think contain takes the image dimensions by default. How is `center` different from it?

> will check really need that timeout and searching workarounds.

Timeout is considered a workaround. And I don't see the need for it. 

> for web version we dont need correction image size actually step 1 can keep function continue after that this.state.zoomScale can be set.

I think your step 1 is directly affected if we don't we step 2. But I didn't understand your meaning here. Could you please explain it?  

Note: There is a lot going on in the imageView component. It also expects the user to zoom the image. But zooming should only be allowed under the image boundary. This means as soon as you move your cursor out of the image, zooming is not available. All of this should work as it is.

**Comment by metehanozyurt** [comment]
First time I test I thought step 1 is fixing problem. But i keep testing repeatedly then realize that not enough.Because functions call order can be change sometimes like this video. Thats why i try to timeout but much wise choice i will think.
   
https://user-images.githubusercontent.com/6093207/162569484-d35970dd-e961-4acf-b215-40ae5b329fe4.mov


![WhatsApp Image 2022-04-09 at 14 41 10](https://user-images.githubusercontent.com/6093207/162572604-0be6564e-0d63-489f-8bd9-c484cdf5366f.jpeg)




According to Image component's documents resizeModes informations like this.

contain:
Scale the image uniformly (maintain the image's aspect ratio) so that both dimensions (width and height) of the image will be equal to or less than the corresponding dimension of the view (minus padding).

center:
Scale the image down so that it is completely visible, if bigger than the area of the view. The image will not be scaled up.
 
"contain" scale up images , "center" not scale up small images. On mWeb using center on small images can fix problem for scale.

**Comment by parasharrajat** [comment]
> First time I test I thought step 1 is fixing problem. But i keep testing repeatedly then realize that not enough.Because functions call order can be change sometimes like this video. Thats why i try to timeout but much wise choice i will think

Yup, this is the issue. So do you consider setTimeout a solution to this? What else can be done to have a predefined order?

--- 

Waiting on some internal discussion.

**Comment by metehanozyurt** [comment]
I've been thinking about solving this problem for a while and feel bad about setTimeout solution. Instead of using setTimeout, I have developed a solution proposal if you want to examine it @parasharrajat 🙏 .

We can forget about setTimeout solution.  

If we change this line works fine. I do this for native version but i skip this for web version. This is scale images height and width  when setImageRegion calls last.

https://github.com/Expensify/App/blob/9a8291ba97c381b651b5344d6fa471c51ae70336/src/components/ImageView/index.js#L144-L147

to this
```js
            //if (aspectRatio > 1) { 
            //    width *= (aspectRatio); 
            //   height *= (aspectRatio); 
            //} 
```
If we arrange it this way, there is no problem. Sorry for not finding a solution sooner.😟

**Comment by parasharrajat** [comment]
Ok. Thanks for bringing this up. I think we want to keep a minimum height and width for the image. If the user uploads a very small image of 2x2px then it would be barely visible. Let me discuss this on slack first.

**Comment by parasharrajat** [comment]
Ok, I think we are good here with the expected behavior.  

@metehanozyurt 's proposal looks good to me. Rest we can test on PR. 

cc: @flodnv 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by flodnv** [comment]
I'm unclear: is the final proposal what is in [this comment](https://github.com/Expensify/App/issues/8471#issuecomment-1092091939)? It seems we have made other decisions since then?

**Comment by parasharrajat** [comment]
Sorry, yeah post is confusing. @metehanozyurt Would you mind writing the whole solution in a single post for native and web?

**Comment by flodnv** [comment]
Yes please, and please clarify why we need platform specific code (I did not really follow that part)

**Comment by flodnv** [comment]
Thanks! Sounds good 👍 @kevinksullivan please hire @metehanozyurt

**Comment by kevinksullivan** [comment]
Hired @parasharrajat for c+. @metehanozyurt please apply to the job! 

https://www.upwork.com/jobs/~01079bbed339023db8

**Comment by metehanozyurt** [comment]
Thank you @flodnv , @parasharrajat , @kevinksullivan . I am preparing my PR with your permission.

**Comment by kevinksullivan** [comment]
Thanks @metehanozyurt , please remember to apply to the job so i can make the payment when this is complete.

**Comment by flodnv** [comment]
PR is nearing the final review stage 👌

**Comment by flodnv** [comment]
PR is in review.

**Comment by parasharrajat** [comment]
PR awaiting review from @flodnv.

**Comment by kevinksullivan** [comment]
Paid @parasharrajat and @metehanozyurt . Thanks!
